### PR TITLE
CVSL-1793 add functionality to time out licences on sentence date changes

### PR DESCRIPTION
**Pull request by gurps-singh**
This PR is to add logic to time out licences when a sentence date change pushes a licence into the hard stop period.

### CODE REVIEW

**Review by SebNegahbanMoJ — APPROVED** [review]

File: src/main/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/service/UpdateSentenceDateService.kt [hunk 1]
```diff
@@ -151,6 +157,32 @@ class UpdateSentenceDateService(
     )
   }
 
+  private fun Licence.recordAuditEvent(auditEventDescription: String) {
```
Comment (andrewrlee) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/service/UpdateSentenceDateService.kt:160]: it's a bit weird asking for a licence entity to do something that talks to the outside world.
Entities should generally be treated as individual data objects with behaviour that modifies its own state. 
 
You could use a `with` inside a separate method to make accessing fields easier:
```kt
  with(licence) {
auditEventRepository.saveAndFlush(
      AuditEvent(
        licenceId = this.id,
        username = "SYSTEM",
        fullName = "SYSTEM",
        eventType = AuditEventType.SYSTEM_EVENT,
        summary = "$auditEventDescription for ${this.forename} ${this.surname}",
        detail = "ID ${this.id} type ${this.typeCode} status ${this.statusCode} version ${this.version}",
      ),
    )
  }
```  
which would have the same affect to put the licence in scope

File: src/main/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/service/UpdateSentenceDateService.kt [hunk 2]
```diff
@@ -51,18 +57,18 @@ class UpdateSentenceDateService(
...
-      ),
-    )
+    val licenceIsNowInHardStopPeriod = releaseDateService.isInHardStopPeriod(updatedLicenceEntity)
+
+    if (licenceIsNowInHardStopPeriod && updatedLicenceEntity is CrdLicence) {
+      val timedOutLicence = updatedLicenceEntity.timeOut()
+      licenceRepository.saveAndFlush(timedOutLicence)
+      updatedLicenceEntity.recordAuditEvent("Sentence dates updated")
+      timedOutLicence.recordAuditEvent("Licence automatically timed out")
+      timedOutLicence.recordLicenceEvent("Licence automatically timed out")
```
Comment (andrewrlee) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/service/UpdateSentenceDateService.kt:67]:
```
updatedLicenceEntity.recordAuditEvent("Sentence dates updated")
timedOutLicence.recordAuditEvent("Licence automatically timed out")
```

these two licences are the "same" (share the same id) - would just refer to the updated one after it's been updated. (the other one can be seen as a stale copy)

File: src/main/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/service/UpdateSentenceDateService.kt (hunk 1 above)
Comment (gurps-singh) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/service/UpdateSentenceDateService.kt:160]: Done

File: src/main/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/service/UpdateSentenceDateService.kt (hunk 2 above)
Comment (gurps-singh) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/service/UpdateSentenceDateService.kt:67]: Done

**Review by andrewrlee — APPROVED** [review]
LGTM

**Pull request merged**